Calculate primitive's AABB correctly

daemyung opened the pull request:
AABB of morph target is displacement data so it should be added to primitive's base AABB.

Fix #6231

daemyung commented:
You can get more context at [here](https://github.com/zeux/meshoptimizer/pull/504).

pixelflinger commented:
@daemyung so how do you solve the problem in the end?

daemyung commented:
> @daemyung so how do you solve the problem in the end?

@pixelflinger The solution is just using primitive's AABB. So I deleted code that merging morph targets' AABB.

pixelflinger commented:
> > @daemyung so how do you solve the problem in the end?
> 
> @pixelflinger The solution is just using primitive's AABB. So I deleted code that merging morph targets' AABB.

I guess what I don't understand is why the AABB are correct if you only use the primitive's AABB? we're saying that it is the responsibility of the asset creator to set  primitive AABB that will work with all weights (like we do for skinning)? I'm okay with that, just want to understand.

daemyung commented:
> I guess what I don't understand is why the AABB are correct if you only use the primitive's AABB? 

Technically it's wrong if we just use primitive's base AABB. Application needs to calculate AABB with morphing. But the problem is that glTF spec doesn't define the range of morph target weight so we can't get the size of model well. For example morph target weight could be 1.0f or 10000.0f. So filament needs to define the range of morph target weight then we can use morph target's AABB. For example [three.js](https://github.com/mrdoob/three.js/blob/868aee4c3bda29077ba53989cea07948a474782e/examples/js/loaders/GLTFLoader.js#L3905) defines this range from -1.0 to 1.0. But filament doesn't have any spec for this so I look into [recomputeBoundingBoxes](https://github.com/google/filament/blob/15ba54c4c51fcf1ff80790f4223984e5db0106cb/libs/gltfio/include/gltfio/FilamentInstance.h#L138) in gltfio and filament doesn't use morph target to calculate model's AABB. So I change code like this.

> it is the responsibility of the asset creator to set primitive AABB that will work with all weights (like we do for skinning)?

I think so too if animations for morphing is defined. Without animations for morphing we can't know the range of morph target weight.

daemyung commented:
@pixelflinger After talking with you, I have changed my mind. Updated code is calculating AABB with morph target's AABB.

<img width="356" alt="image" src="https://user-images.githubusercontent.com/7459074/198562801-b97eeea1-fdb7-40bd-9a04-577b15fc0cc2.png">

romainguy commented on libs/gltfio/src/AssetLoader.cpp:
```diff
@@ -982,10 +983,20 @@ bool FAssetLoader::createPrimitive(const cgltf_primitive& inPrim, Primitive* out
-            outPrim->aabb.max = max(outPrim->aabb.max, float3(maxp[0], maxp[1], maxp[2]));
+
+            // We assume that the range of morph target weight is [-1,1].
+            targetAabb.min += float3(minp[0], minp[1], minp[2]);
```
Why are we adding the mins here?

daemyung replied:
I guess morph weight could be -1. So model size could increase opposite direction.

romainguy replied:
I don't see how adding the mins makes sense. This can grow the aabb?

daemyung replied:
I didn't check. But this AABB is displacement data (deltas) so it could be I guess. For example base primitive AABB min is [0, 0, 0] and target AABB(for delta) min is [-1, -1, -1] then final AABB min will be [-1, -1, -1].

romainguy replied:
Ah If it's displacement then it makes more sense.

pixelflinger commented:
> @pixelflinger After talking with you, I have changed my mind. Updated code is calculating AABB with morph target's AABB assuming range is -1 to 1. Below figure is the result.

I think this makes sense because you can always author the data so it does what you need with weights between -1 and 1.

We should probably document that somewhere. Like saying we allow weights out of the [-1,1]  range, but that bounding boxes in gltfio assume that.
